### ISSUE TITLE
BatchNorm changes the data type.

### ISSUE THREAD

**Issue by ayush1999** [FluxML/Flux.jl#260]
The BatchNorm operation is changing the data type. For example:
```

>>> a = convert(Array{Float32,4}, rand(5,5,1,1))    # Tensor of Float32.
>>> a = Tracker.TrackedArray(a)                     # TrackedArray of Float32
>>> b = BatchNorm(1, ϵ=1e-5, momentum = 0.9)
>>> b(a)
Tracked 5×5×1×1 Array{Float64,4}:
[:, :, 1, 1] =
 -0.353323  -1.22248     0.219534  -0.146294   -1.4266   
  1.39974    1.07692     0.925905   0.0425806   0.218777 
  0.920126  -1.55046     1.44969   -1.16887     1.06475  
 -0.143462   1.37654    -0.47538   -1.81898    -0.245258 
 -1.58401   -0.0710794   1.10551    0.388538    0.0175963

```

As we can see, the data changed from tensor of Float32 to tensor of Float64. As a result, performing further computation isn't possible (because the weights and inputs need to have same type)

**Comment by ayush1999** [comment]
@MikeInnes This is interfering with onnx work, could you please have a look?

**Comment by CarloLucibello** [comment]
Unfortunately, this is a case of #231:  `BN.γ` and  `BN.β` defaults initialization is to Float64 arrays. I'm not sure that #261, i.e. converting the final output to the input data,  would be a proper solution for this. One could argue that we would need such a convert for the output of any layer (Dense, Conv, ....)

A verbose workaround for this issue is to use pass the arguments `initβ=x->zeros(Float32, x), initγ=x->ones(Float32, x)` to the BatchNorm constructor. 

More conveniently and to generally fix #231, we should add a conversion method for the whole model, something like, `model = model |> float32`  and/or add to each constructor an eltype argument.

**Comment by KristofferC** [comment]
You just need to initialize `ϵ` and `momentum` consistently.

```
b = BatchNorm(1, ϵ=Float32(1e-5), momentum = Float32(0.9))
```
